enhancements/scheduling/prevent-user-workloads-from-being-scheduled-on-control-plane-nodes.md

```diff
@@ -0,0 +1,412 @@
+
+- Protect Specialized Resources: Restrict scheduling of regular user workloads on nodes with specialized resources (e.g., GPU-enabled nodes) to ensure these costly resources are reserved for appropriate workloads.
+
+- Emergency Override Capability: Allow administrators to temporarily override scheduling restrictions for emergency or maintenance tasks, ensuring critical operations can be performed when necessary without compromising the cluster's overall security posture.
+
+### Non-Goals
+
+<!-- What is out of scope for this proposal? Listing non-goals helps to
+focus discussion and make progress. Highlight anything that is being
+deferred to a later phase of implementation that may call for its own
+enhancement. -->
+
+- Granular Pod-Level Scheduling Controls: The enhancement will not introduce fine-grained controls for individual pod scheduling decisions beyond the existing Kubernetes mechanisms (e.g., taints, tolerations, and affinity rules).
+
+- Automated User or Service Account Management: Automatically managing or updating the list of service accounts or namespaces authorized to schedule workloads on control plane nodes is out of scope. This process remains a manual administrative responsibility.
+
+- Real-Time Resource Allocation Optimization: The proposal does not aim to dynamically optimize resource allocation or scheduling decisions based on real-time cluster utilization or performance metrics.
+
+## Proposal
+
+<!-- This is where we get down to the nitty gritty of what the proposal
+actually is. Describe clearly what will be changed, including all of
+the components that need to be modified and how they will be
+different. Include the reason for each choice in the design and
+implementation that is proposed here, and expand on reasons for not
+choosing alternatives in the Alternatives section at the end of the
+document. -->
+
+### Overview
+This enhancement proposes the implementation of a more robust and flexible mechanism for enforcing pod placement rejections on OpenShift clusters. It will focus on preventing user workloads from being scheduled on control plane nodes and allowing for exceptions based on administrative configurations.
+
+### Implementation Strategies
+
+Update OpenShift Components for Taint Tolerance: Modify OpenShift operators and control plane components to include tolerations for the NoExecute taint on control plane nodes that admins can apply to achieve the goals described here. This will ensure that essential services and components are not evicted or prevented from scheduling on these nodes due to the taint. The update process should involve a thorough review of all default and critical components to add the necessary toleration, ensuring they continue to operate as expected in environments where the NoExecute taint is applied. This step is crucial for maintaining cluster stability and ensuring that core functionalities are not disrupted by the enforcement of the new scheduling policies. 
```

Comment:
> are not evicted or prevented from scheduling on these nodes due to the taint.

are not evicted or prevented from running on these nodes due to the taint. # scheduling -> running

Reply:
✅